junit/src/main/java/org/kaazing/k3po/junit/rules/ScriptRunner.java

```diff
@@ -214,6 +214,7 @@ public void notifyBarrier(final String barrierName) throws InterruptedException
 
 
     public void dispose() throws Exception {
+        latch.awaitFinished();
```

Comment:
We need to fix all the cases where latch.awaitFinished() cause test blocking.  John and Jitu hit a case recently in debugging, and have or will be filing a bug.

We can chat in person, but there are use cases to order the rules in various orders, so I would suggest not trying to solve for what we consider common, but try to solve for all.

Your reasoning on the bug and finding it makes complete sense.  Awesome job! but I think we can implement the fix better as it adds complexity and fatal error condition where we block forever.  I would suggest trying to fix it where you observed it,  I think you are talking about this code block:

```java
            } catch (AssumptionViolatedException e) {

                if (!latch.isFinished()) {
                    scriptRunner.abort();
                }

                throw e;
            } 
...
        } finally {
            try {
                scriptRunner.dispose();
```

The catch code should probably wait for an amount of time for the finish (see the catch directly below), i.e.

```
                    try {
                        // wait at most 5sec for the observed script (due to the abort case)
                        // should take less than a second for K3PO to complete
                        ScriptPair scripts = scriptFuture.get(5, SECONDS);

                        try {
                            assertEquals("Specified behavior did not match", scripts.getExpectedScript(),
                                    scripts.getObservedScript());
                            // Throw the original exception if we are equal
                            throw cause;
                        } catch (ComparisonFailure f) {
                            // throw an exception that highlights the difference in behavior, but caused by the timeout
                            // (or
                            // original exception)
                            f.initCause(cause);
                            throw f;
                        }
```

Making the two different catch the exact same and allowing us to not special case them (simplification!!).  I haven't tried this and this is shoot from the hip suggestion, you might have to rework it.